# 📌 Numerical Problems



---
## **Exhaustive Enumeration**

### 3️⃣ Find the cube root


In [ ]:
import time 
x = int(input("Enter an integer: "))
start = time.time()
ans = 0
while ans**3 < abs(x):
    ans += 1
if ans**3 != abs(x):
    print(f'{x} is not a perfect cube')
else:
    if x < 0:
        ans = -ans
    print(f'{ans} is the cuberoot of {x}')
end = time.time()
print(f'total time {(end-start):0.4f}')

1949306 is the cuberoot of 7406961012236344616
total time 0.2480


* The algorithmic technique used in this program is a variant of guess-and-check called exhaustive enumeration

* We enumerate all possibilities until we get to the right answer or exhaust the space of possibilities

* Exhaustive enumeration algorithms are often the most practical way to solve a problem

* They are typically easy to implement and easy to understand. And, in many cases, they run fast enough for all
practical purposes

* Try 7406961012236344616.

* Modern computers are amazingly fast.It takes less than one nanosecond—one billionth of a second—to execute an instruction

* For perspective, it takes slightly more than a nanosecond for light to travel a single foot (0.3 meters)

### 4️⃣ Method to find prime number

The simplest way to find out if an integer, x, greater than 3 is prime, is to divide x by each integer between 2 and, x-1. If the
remainder of any of those divisions is 0, x is not prime, otherwise x is prime

In [32]:
num = int(input("Enter the integer greater than 2:"))
smallest_divisor = None
for guess in range(2,num):
    if num % guess == 0:
        smallest_divisor = guess
        break
if smallest_divisor is None:
    print(f'The number {num} is a prime number')
else:
    print(f'The smallest divisor is {smallest_divisor}')

The number 31 is a prime number


In [37]:
num = int(input("Enter the integer greater than 2"))
smallest_divisor = None 

if num % 2 == 0:
    smallest_divisor = 2
else:
    for guess in range(3,num,2):
        if num % guess == 0:
            smallest_divisor = guess
            break 
if smallest_divisor is None: 
    print("The number is prime")
else:
    print(f"The smallest divisor is {smallest_divisor}")


The number is prime


### Exercise 5️⃣ 
Write a program that asks the user to enter an integer and prints two integers, root and pwr, such that 1 < pwr < 6 and root**pwr is equal to the integer entered by the user. If no such pair of integers exists, it should print a message to that effect.

In [20]:
num = int(input("Enter an integer: "))
isFound = False
for root in range(1,num):
    for pwr in range(2,6):
        if root**pwr >= num:
            if  root**pwr == num: 
                isFound = True
            break
    if isFound :
        break
if isFound and root**pwr == num:
    print(f"The number = {root}^{pwr}")
if root**pwr != num:
    print("No such pair exists")
      
            

The number = 15^2


## **Approximate Solutions**

* what should the program do if asked to find the square root of 2? 

    * The square root of 2 is not a rational number. This means that there is no way to precisely represent its value as a finite string of digits (or as a float), so the problem as initially stated cannot be solved.

    * But we can find an approximation to the square root : a solution which is "close enough" to the actual solution

* What is this close enough ?

#### 6️⃣ Algorithm for the square root of x ?



In [ ]:
x = 0.25
epsilon = 0.01
step = epsilon**2
num_guesses = 0
ans = 0.0 
while abs(ans**2 - x) >= epsilon and ans <= x:
    ans += step 
    num_guesses += 1
if abs(ans**2 - x) < epsilon:
    print(f'The approximate square root of {x} is {ans} after {num_guesses} guessess')
else:
    print(f'The method fails to find a solution after {num_guesses} guessess')

The method fails to find a solution after 2501 guessess


In [6]:
x = int(input("Enter an integer: "))
epsilon = 0.01
step = epsilon**3
num_guesses = 0
ans = 0.0 
while abs(ans**2 - x) >= epsilon and ans*ans <= x:
    ans += step 
    num_guesses += 1
if abs(ans**2 - x) < epsilon:
    print(f'The approximate square root of {x} is {ans} after {num_guesses} guessess')
else:
    print(f'The method fails to find a solution after {num_guesses} guessess')

The approximate square root of 123456 is 351.36304620491023 after 351363047 guessess


* The number of iterations depends upon how close the answer is to our starting point and the size of the steps

* The program will execute the loop atmost $\frac{x}{step}$ times

# **Searching for a Word in a Hard-Copy Dictionary**

## **Introduction**
Consider the problem of discovering whether a word starting with a given sequence of letters appears in a hard-copy dictionary of the English language. 

One possible approach is **exhaustive enumeration**, where we start at the first word and examine each word sequentially until we either find a match or reach the end of the dictionary.

## **Exhaustive Enumeration and Search Complexity**
- If the dictionary contains **$ n$ words**, we perform a **linear search**.
- If the word exists in the dictionary, we will find it at some position $ k $ where $ k $ is uniformly distributed between $ 1 $ and $ n $.
- On average, the search will take ** $ n/2 $ probes** before finding the word.
- If the word is **not present**, we will have to check all ** $n$ words**, leading to a worst-case of  $n $ probes.

## **Mathematical Justification of $ n/2 $**
Let $X $ be the position at which we find the word. Since the word can be at any position **uniformly** at random:

$$
P(X = k) = \frac{1}{n}, \quad \text{for } k = 1, 2, ..., n.
$$

The expected number of probes required to find the word is:

$$
E[X] = \sum_{k=1}^{n} k \cdot P(X = k)
$$

Substituting $ P(X = k) = \frac{1}{n} $:

$$
E[X] = \frac{1}{n} \sum_{k=1}^{n} k
$$

Using the formula for the sum of the first \( n \) natural numbers:

$$
\sum_{k=1}^{n} k = \frac{n(n+1)}{2}
$$

we get:

$$
E[X] = \frac{1}{n} \times \frac{n(n+1)}{2} = \frac{n+1}{2}
$$

For large \( n \), this simplifies to:

$$
E[X] \approx \frac{n}{2}
$$

Thus, on average, we need to examine **$ n/2 $ words** before finding the desired word.



## **Worst-Case Analysis**
If the word is **not present in the dictionary**, we must check **all $ n $ words** before concluding that it does not exist.

Thus, the worst-case scenario requires **$ n $ probes**.

## **Real-Life Intuition**
Those who have had the pleasure of looking up a word in a **physical dictionary** rather than an online one would never proceed in this exhaustive manner. Instead, we typically:
- Use **alphabetical ordering** to jump to an approximate location.
- Use **binary search**-like techniques by flipping to a section and refining our search.
- This approach significantly reduces the number of probes needed.
- Using the same idea for finding the square root of $x$


___
# **Bisection Search**

* An example of sucessive approximation method


In [2]:
x = int(input("Enter an integer: "))
epsilon = 0.01
num_guesses = 1
low = 0
high = max(1.0, x)
ans = (high + low) / 2.0
while abs(ans**2 - x) >= epsilon:
    print(f'low:{low}, high:{high}, ans:{ans}, guess:{num_guesses}')
    if ans**2 > x:
        high = ans 
    else:
        low = ans
    ans = (high+low)/2
    num_guesses += 1

ValueError: invalid literal for int() with base 10: ''


## **Introduction**
The term **"bisection"** comes from the **Latin word "bi-" (meaning "two")** and **"section" (meaning "cut")**. 

In **bisection search**, the search space is **cut in half (divided into two parts) at each step**, which is why the name **bisection** is used.

### **1. Start with an Interval**
- Suppose we want to find a number \( y \) such that $$ f(y) = 0 $$ in an interval $$ [a, b] $$.

### **2. Find the Midpoint**
- Compute the midpoint:

  $$
  m = \frac{a + b}{2}
  $$

### **3. Decide Which Half to Keep**
- If $$ f(m) $$ is close enough to zero, we stop.
- If $$ f(m) $$ has the **same sign** as $$ f(a) $$, then the root must be in $$ [m, b] $$, so we **cut away the left half** and update $$ a = m $$.
- Otherwise, the root must be in $$ [a, m] $$, so we **cut away the right half** and update $$ b = m $$.

### **4. Repeat Until Convergence**
- Each step **bisects the interval**, reducing its size by half.
- The search space shrinks **exponentially**, ensuring convergence.

# **Proof: Bisection Search Shrinks the Search Space Exponentially**

## **1. Defining the Problem**
Suppose we are searching for a value (e.g., a root of a function) in an **initial search interval**:

$$
[a_0, b_0]
$$

where:

$$
a_0 \text{ is the lower bound,}
$$

$$
b_0 \text{ is the upper bound,}
$$

$$
L_0 = b_0 - a_0 \text{ is the initial interval length.}
$$

At each step, we compute the **midpoint**:

$$
m_k = \frac{a_k + b_k}{2}
$$

and decide which half to keep based on the function's sign.

---

## **2. How the Interval Shrinks**
In **each iteration**, the new interval **halves**:
- If the root is in the left half, update:

  $$
  b_{k+1} = m_k
  $$

- If the root is in the right half, update:

  $$
  a_{k+1} = m_k
  $$

The **length of the interval** after $ k $ iterations is:

$$
L_k = \frac{L_0}{2^k}
$$

where $ L_0$ is the initial length of the interval.

This shows that **the interval length decreases exponentially** with base 2.

---

## **3. Exponential Decay**
From the formula:

$$
L_k = \frac{L_0}{2^k}
$$

- The search space shrinks **exponentially** because:

  $$
  2^k \text{ grows exponentially.}
  $$

- After $k$ iterations, the remaining search space is reduced to:

  $$
  \frac{1}{2^k} \text{ of the original space.}
  $$

---

## **4. Number of Iterations Needed for Convergence**
Suppose we want to stop when the interval size is smaller than a small error tolerance $ \epsilon$, meaning:

$$
L_k < \epsilon
$$

Using the shrinking formula:

$$
\frac{L_0}{2^k} < \epsilon
$$

Solving for $ k $:

$$
k > \log_2 \left(\frac{L_0}{\epsilon}\right)
$$

Thus, the number of iterations required to achieve accuracy $ \epsilon$ is:

$$
k = \lceil \log_2 (L_0 / \epsilon) \rceil
$$

This confirms **exponential convergence**, as the number of iterations is proportional to the logarithm of the precision.

---

## **Conclusion**
- **The search space shrinks exponentially** as:

  $$
  L_k = \frac{L_0}{2^k}
  $$

- **The number of iterations needed for accuracy**  $\epsilon$ **is logarithmic** in the initial interval size.
- **Bisection search is very efficient**, requiring at most:

  $$
  O(\log_2 (L_0 / \epsilon))
  $$

  iterations.

🚀 This is why **bisection search converges quickly** compared to linear search methods!

#### **Exercise-7** 
The Empire State Building is 102 stories high. A
man wanted to know the highest floor from which he could drop an
egg without the egg breaking. He proposed to drop an egg from the
top floor. If it broke, he would go down a floor, and try it again. He
would do this until the egg did not break. At worst, this method
requires 102 eggs. Implement a method that at worst uses seven
eggs.

In [ ]:
low = 0
high = 102
# ans = (low+high)//2
num_guesses = 0
while num_guesses < 7:
    ans = (low+high)//2
    num_guesses += 1
    # print(f'high:{high},low:{low},guess:{num_guesses},ans:{ans}')
    status = input(f'Did the egg broke from the floor {ans} ?')
    if status.lower() == 'yes':
        high = ans
    elif status.lower() == 'no':
        low = ans 
    else:
        print(f'Not a valid input')
    
print(low)

# Floating Point Numbers and its Pitfalls

9


## Floating Points and it Quirkness

In [1]:
x = 0.0
for i in range(10):
    x = x + 0.1
    if x == 1.0:
        print(x,'=1.0')
    else:
        print(x,'is not 1.0')


0.1 is not 1.0
0.2 is not 1.0
0.30000000000000004 is not 1.0
0.4 is not 1.0
0.5 is not 1.0
0.6 is not 1.0
0.7 is not 1.0
0.7999999999999999 is not 1.0
0.8999999999999999 is not 1.0
0.9999999999999999 is not 1.0
